source/hp/refinement.cc

```diff
@@ -843,6 +843,97 @@ namespace hp
       // - always raise levels to match criterion, never lower them
       // - exchange level indices on ghost cells
 
+      // Function that updates the level of neighbor to fulfill difference
+      // criterion, and returns whether it was changed.
+      const auto update_neighbor_level =
+        [&future_levels, max_difference](const auto &     neighbor,
+                                         const level_type cell_level) -> bool {
+        Assert(neighbor->is_active(), ExcInternalError());
+        // We only care about locally owned neighbors. If neighbor is a ghost
+        // cell, its future FE index will be updated on the owning process and
+        // communicated at the next loop iteration.
+        if (neighbor->is_locally_owned())
+          {
+            const level_type neighbor_level = static_cast<level_type>(
+              future_levels[neighbor->global_active_cell_index()]);
+
+            // ignore neighbors that are not part of the hierarchy
+            if (neighbor_level == invalid_level)
+              return false;
+
+            if ((cell_level - max_difference) > neighbor_level)
+              {
+                future_levels[neighbor->global_active_cell_index()] =
+                  cell_level - max_difference;
+
+                return true;
+              }
+          }
+
+        return false;
+      };
+
+      // For cells to be h-coarsened, we need to determine a future FE for the
+      // parent cell, which will be the dominated FE among all children
+      // However, if we want to enforce the max_difference criterion on all
+      // cells on the updated mesh, we will need to simulate the updated mesh on
+      // the current mesh.
+      //
+      // As we are working on p-levels, we will set all siblings that will be
+      // coarsened to the highest p-level among them. The parent cell will be
+      // assigned exactly this level in form of the corresponding FE index in
+      // the adaptation process in
+      // Triangulation::execute_coarsening_and_refinement().
+      //
+      // This function takes a cell and sets all its siblings to the highest
+      // p-level among them. Returns whether any future levels have been
+      // changed.
+      const auto prepare_level_for_parent = [&](const auto &neighbor) -> bool {
+        Assert(neighbor->is_active(), ExcInternalError());
+        if (neighbor->coarsen_flag_set() && neighbor->is_locally_owned())
+          {
+            const auto parent = neighbor->parent();
+
+            std::set<unsigned int> future_levels_children;
```

Comment:
Why not use here a vector?

Reply:
I access the last element of the `set` to determine the highest value.
I could also use `vector` and use `max_element` instead. I don't know what's more practical. Any opinions?

Reply:
Vector :)